service/matching/matching_engine_test.go

```diff
@@ -2322,6 +2322,124 @@ func (s *matchingEngineSuite) TestAddActivityTask_ForVersionedWorkflows_Silently
 	s.Require().NoError(err)
 }
 
+func (s *matchingEngineSuite) TestPersistUnknownBuildId_Poll() {
+	namespaceId := namespace.ID(uuid.New())
+	tl := "makeToast"
+	tlID := newTestTaskQueueID(namespaceId, tl, enumspb.TASK_QUEUE_TYPE_WORKFLOW)
+
+	scope := tally.NewTestScope("test", nil)
+	s.matchingEngine.metricsHandler = metrics.NewTallyMetricsHandler(metrics.ClientConfig{}, scope)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Millisecond)
+	defer cancel()
+
+	_, err := s.matchingEngine.getTask(ctx, tlID, normalStickyInfo, &pollMetadata{
+		workerVersionCapabilities: &commonpb.WorkerVersionCapabilities{
+			BuildId:       "unknown",
+			UseVersioning: true,
+		},
+	})
+	s.Error(err) // deadline exceeded or canceled
+
+	unknownCtr := scope.Snapshot().Counters()["test.unknown_build_polls+namespace="+matchingTestNamespace+",operation=TaskQueueMgr,task_type=Workflow,taskqueue=makeToast"]
+	s.Equal(int64(1), unknownCtr.Value())
+}
+
+func (s *matchingEngineSuite) TestPersistUnknownBuildId_Add() {
+	namespaceId := namespace.ID(uuid.New())
+	tl := "makeToast"
+
+	scope := tally.NewTestScope("test", nil)
+	s.matchingEngine.metricsHandler = metrics.NewTallyMetricsHandler(metrics.ClientConfig{}, scope)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Millisecond)
+	defer cancel()
+
+	s.mockMatchingClient.EXPECT().UpdateWorkerBuildIdCompatibility(gomock.Any(), &matchingservice.UpdateWorkerBuildIdCompatibilityRequest{
+		NamespaceId: namespaceId.String(),
+		TaskQueue:   tl,
+		Operation: &matchingservice.UpdateWorkerBuildIdCompatibilityRequest_PersistUnknownBuildId{
+			PersistUnknownBuildId: "unknown",
+		},
+	}).Return(&matchingservice.UpdateWorkerBuildIdCompatibilityResponse{}, nil)
+
+	_, err := s.matchingEngine.AddWorkflowTask(ctx, &matchingservice.AddWorkflowTaskRequest{
+		NamespaceId:            namespaceId.String(),
+		Execution:              &commonpb.WorkflowExecution{RunId: "run", WorkflowId: "wf"},
+		ScheduledEventId:       0,
+		TaskQueue:              &taskqueuepb.TaskQueue{Name: tl, Kind: enumspb.TASK_QUEUE_KIND_NORMAL},
+		ScheduleToStartTimeout: timestamp.DurationFromSeconds(100),
+		ForwardedSource:        "somewhere", // force sync match only
+		VersionDirective: &taskqueue.TaskVersionDirective{
+			Value: &taskqueue.TaskVersionDirective_BuildId{
+				BuildId: "unknown",
+			},
+		},
+	})
+	s.ErrorIs(err, errRemoteSyncMatchFailed)
+
+	unknownCtr := scope.Snapshot().Counters()["test.unknown_build_tasks+namespace="+matchingTestNamespace+",operation=TaskQueueMgr,task_type=Workflow,taskqueue=makeToast"]
+	s.Equal(int64(1), unknownCtr.Value())
```

Comment:
It's really good that we're asserting these oft-forgotten things like error types and metrics. However, shouldn't there also be more critical side effects that we're verifying? For example, shouldn't we verify that something is or is not persisted? Am I right in understanding that there is nothing persisted here due to the error, and that is implicitly asserted due to the absence of an EXPECT() statement?